v2/bundle/spiffebundle/bundle.go

```diff
@@ -3,175 +3,411 @@ package spiffebundle
-	panic("not implemented")
+	if jwtBundle != nil {
+		return &Bundle{
+			trustDomain: jwtBundle.TrustDomain(),
+			jwtKeys:     jwtBundle.JWTKeys(),
+		}
+	}
+	return &Bundle{}
+}
+
+// FromX509Roots creates a bundle from X.509 certificates.
+func FromX509Roots(trustDomain spiffeid.TrustDomain, x509Roots []*x509.Certificate) *Bundle {
+	return &Bundle{
+		trustDomain: trustDomain,
+		x509Roots:   x509Roots,
+	}
+}
+
+// FromJWTKeys creates a new bundle from JWT public keys.
+func FromJWTKeys(trustDomain spiffeid.TrustDomain, jwtKeys map[string]crypto.PublicKey) *Bundle {
+	return &Bundle{
+		trustDomain: trustDomain,
+		jwtKeys:     jwtKeys,
+	}
 }
 
 // TrustDomain returns the trust domain that the bundle belongs to.
 func (b *Bundle) TrustDomain() spiffeid.TrustDomain {
-	panic("not implemented")
+	return b.trustDomain
 }
 
 // X509Roots returns the X.509 roots in the bundle.
 func (b *Bundle) X509Roots() []*x509.Certificate {
-	panic("not implemented")
+	b.mtx.RLock()
+	defer b.mtx.RUnlock()
+
+	return b.x509Roots
 }
 
 // AddX509Root adds an X.509 root to the bundle. If the root already
 // exists in the bundle, the contents of the bundle will remain unchanged.
-func (b *Bundle) AddX509Root(*x509.Certificate) {
-	panic("not implemented")
+func (b *Bundle) AddX509Root(x509Root *x509.Certificate) {
+	b.mtx.Lock()
+	defer b.mtx.Unlock()
+
+	for _, r := range b.x509Roots {
+		if x509util.CertsEqual(r, x509Root) {
+			return
+		}
+	}
+
+	b.x509Roots = append(b.x509Roots, x509Root)
 }
 
 // RemoveX509Root removes an X.509 root from the bundle.
-func (b *Bundle) RemoveX509Root(*x509.Certificate) {
-	panic("not implemented")
+func (b *Bundle) RemoveX509Root(x509Root *x509.Certificate) {
+	b.mtx.Lock()
+	defer b.mtx.Unlock()
+
+	for i, r := range b.x509Roots {
+		if x509util.CertsEqual(r, x509Root) {
+			b.x509Roots = append(b.x509Roots[:i], b.x509Roots[i+1:]...)
+			return
+		}
+	}
 }
 
 // HasX509Root checks if the given X.509 root exists in the bundle.
 func (b *Bundle) HasX509Root(root *x509.Certificate) bool {
-	panic("not implemented")
+	b.mtx.RLock()
+	defer b.mtx.RUnlock()
+
+	for _, r := range b.x509Roots {
+		if x509util.CertsEqual(r, root) {
+			return true
+		}
+	}
+	return false
 }
 
 // JWTKeys returns the JWT keys in the bundle, keyed by key ID.
 func (b *Bundle) JWTKeys() map[string]crypto.PublicKey {
-	panic("not implemented")
+	b.mtx.RLock()
+	defer b.mtx.RUnlock()
+
+	return b.jwtKeys
 }
 
-// FindKey finds the JWT key with the given key id from the bundle. If the key
+// FindJWTKey finds the JWT key with the given key id from the bundle. If the key
 // is found, it is returned and the boolean is true. Otherwise, the returned
 // value is nil and the boolean is false.
 func (b *Bundle) FindJWTKey(keyID string) (crypto.PublicKey, bool) {
-	panic("not implemented")
+	b.mtx.RLock()
+	defer b.mtx.RUnlock()
+
+	if jwtKey, ok := b.jwtKeys[keyID]; ok {
+		return jwtKey, true
+	}
+	return nil, false
```

Comment:
nit: map behavior already returns a zero value for keys that aren't set.
```suggestion
	jwtKey, ok := b.jwtKeys[keyID]
	return jwtKey, ok
```